### PR TITLE
Driver for Linear LTC2983 / LTC2984

### PR DESCRIPTION
**Pull request by rleh**
Driver for [Linear LTC2983 / LTC2984](http://www.linear.com/product/LTC2984) *Multi-Sensor High Accuracy Digital Temperature Measurement System*.

Currently only RTDs (*Resistance Temperature Device*, e.g. Pt100) are supported. Thermocouple, thermistor or temperature diode sensors are not supported.

Tested in hardware with a LTC2984 chip and Pt100.

### CODE REVIEW

**Comment by rleh** [comment]
Also tested in hardware with a LTC2983 chip and (again) Pt100.

image

File: examples/stm32f4_discovery/temperature_ltc2984/main.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+									 xpcc::ltc2984::Configuration::SensorType::Pt100,
+									 xpcc::ltc2984::Configuration::Rtd::RsenseChannel::Ch2_Ch1,
+									 xpcc::ltc2984::Configuration::Rtd::Wires::Wire4,
+									 xpcc::ltc2984::Configuration::Rtd::ExcitationMode::Rotation_Sharing,
+									 xpcc::ltc2984::Configuration::Rtd::ExcitationCurrent::Current_500uA,
+									 xpcc::ltc2984::Configuration::Rtd::RtdCurve::European
+									 )));
+		tempSensor.enableChannel(xpcc::ltc2984::Configuration::MuxChannel::Ch4);
+		tempSensor.setChannels();
+		PT_CALL(tempSensor.setChannels());
```
Comment (salkinium) [inline examples/stm32f4_discovery/temperature_ltc2984/main.cpp]: Why is it called twice and without `PT_CALL`?

File: examples/stm32f4_discovery/temperature_ltc2984/main.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+			// we wait until the conversations are done
+			while (PT_CALL(tempSensor.isBusy()))
+			{
+			}
+			logger << "Temperature measurement finished." << xpcc::endl;
+
+			PT_CALL(tempSensor.readChannel(xpcc::ltc2984::Channel::Ch4, temp));
+			logger << "Temperature: " << temp << xpcc::endl;
+
+			logger << "Time: " << (xpcc::Clock::now() - stamp) << xpcc::endl;
```
Comment (salkinium) [inline examples/stm32f4_discovery/temperature_ltc2984/main.cpp:112]: This is really not going to be accurate with printing in between. Then again, it's only 1ms resolution anyways. You could use `xpcc::cortex::CycleCounter::getCount()` to get a much more accurate measurement if you need it.

File: src/xpcc/driver/temperature/ltc2984.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ROBOTERCLUB AACHEN E.V. ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ROBOTERCLUB AACHEN E.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984.cpp]: Please use the short license text as usual.

File: src/xpcc/driver/temperature/ltc2984.hpp [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+			}
+			inline bool adcHardFault() const {
+				return (statusByte & 0b01000000);
+			}
+			inline bool sensorHardFault() const {
+				return (statusByte & 0b10000000);
+			}
+		private:
+			const uint8_t statusByte;
+		};
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984.hpp]: Why is this not a `xpcc::Register`?

File: src/xpcc/driver/temperature/ltc2984.hpp [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Roboterclub Aachen e.V. nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ROBOTERCLUB AACHEN E.V. ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ROBOTERCLUB AACHEN E.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984.hpp]: same here, short license pls.

File: src/xpcc/driver/temperature/ltc2984.hpp [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC__LTC2984_HPP
+#define XPCC__LTC2984_HPP
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984.hpp]: Single `_` please.

File: src/xpcc/driver/temperature/ltc2984.hpp [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+		Ch16 = 15,
+		Ch17 = 16,
+		Ch18 = 17,
+		Ch19 = 18,
+		Ch20 = 19,
+	};
+
+	struct Configuration
+	{
+		enum class TemperatureUnit : uint8_t {
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984.hpp]: Here and below: put `{` on next line.

File: src/xpcc/driver/temperature/ltc2984_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC__LTC2984_HPP
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984_impl.hpp]: single `_`

File: src/xpcc/driver/temperature/ltc2984_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+template < class SpiMaster, class Cs >
+xpcc::ResumableResult<void>
+xpcc::Ltc2984<SpiMaster, Cs>::configure(ltc2984::Configuration::Rejection rejection,
+                                        ltc2984::Configuration::TemperatureUnit temperatureUnit,
+                                        uint8_t muxDelay)
+{
+	RF_BEGIN();
+	static uint8_t config = (static_cast<uint8_t>(rejection) | static_cast<uint8_t>(temperatureUnit));
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984_impl.hpp]: Why can't you use `buffer` here? This adds another byte to the class.

File: src/xpcc/driver/temperature/ltc2984_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+	RF_CALL(writeData(ltc2984::Register::GlobalConfiguration, &config, 1));
+	RF_END_RETURN_CALL(writeData(ltc2984::Register::MuxConfigDelay, muxDelay, 1));
+}
+
+template < class SpiMaster, class Cs >
+xpcc::ResumableResult<bool>
+xpcc::Ltc2984<SpiMaster, Cs>::isBusy()
+{
+	static CommandStatus status;
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984_impl.hpp]: same here as above.

**Review by salkinium — CHANGES_REQUESTED** [review]

File: examples/stm32f4_discovery/temperature_ltc2984/main.cpp (hunk 2 above)
Comment (rleh) [reply examples/stm32f4_discovery/temperature_ltc2984/main.cpp:112]: A conversion takes about 150ms, so 1ms accuracy should be sufficient

File: src/xpcc/driver/temperature/ltc2984.hpp [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+#ifndef XPCC_LTC2984_HPP
+#define XPCC_LTC2984_HPP
+
+#include <xpcc/architecture/interface/spi_device.hpp>
+#include <xpcc/processing/resumable.hpp>
+
+namespace xpcc
+{
+
+struct ltc2984 {
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984.hpp]: 🚨 brace police 🚨

File: src/xpcc/driver/temperature/ltc2984_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+ * The file is part of the xpcc library and is released under the 3-clause BSD
+ * license. See the file `LICENSE` for the full license governing this code.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC_LTC2984_HPP
+	#error	"Don't include this file directly, use 'ltc2984.hpp' instead!"
+#endif
+
+#include "ltc2984.hpp"
```
Comment (salkinium) [inline src/xpcc/driver/temperature/ltc2984_impl.hpp]: nope.

**Review by salkinium — CHANGES_REQUESTED** [review]
Nice driver, just two small changes please.

File: src/xpcc/driver/temperature/ltc2984.hpp (hunk 11 above)
Comment (rleh) [reply src/xpcc/driver/temperature/ltc2984.hpp]: Fixed

**Review by salkinium — DISMISSED** [review]
👍

**Comment by salkinium** [comment]
It looks like I can't rebase this on develop myself, so please rebase this branch.

**Comment by rleh** [comment]
Rebased on develop.

Hmm, *[ ] Allow edits from maintainers.* is not checked, but I don't know why.

**Pull request merged**